### PR TITLE
"Informal" string representation for Layer Class

### PR DESCRIPTION
**Pull request by sidps**
Based on [this mailing list post](https://groups.google.com/forum/#!searchin/lasagne-users/__repr__/lasagne-users/CtDfEC3wVnk/ahS2TefZt0cJ), I wrote a simple __str__ method for the base class which should return an easy to understand "informal" representation of the entire network contained under the specific Layer, similar to what Torch does.

For example, if a model such as the one in mnist.py example is built using

    >>> network = build_model(100, 10)

then one could

    >>> print(network)

to get

    input --> [0] --> [1] --> [2] --> [3] --> [4] --> [5] --> output 
    Details:
            [0]: InputLayer
            [1]: DenseLayer
            [2]: DropoutLayer
            [3]: DenseLayer
            [4]: DropoutLayer
            [5]: DenseLayer

To return more details for each layer, the corresponding layer class (that inherits from Layer) must override a method called __layer_str()

Note that this currently supports only completely sequential networks, but should be simple enough to extend.

### CODE REVIEW

**Comment by ebenolson** [comment]
While a tidier `__repr__` for Layers would be nice, I think this would fit better as a utility function, or as a method of a hypothetical network class. 

I'm also a bit skeptical that this would be simple to extend to branched networks.

**Comment by sidps** [comment]
That does seem like a better idea, I could certainly move this into a util function. However, any detailed representation of a layer would require implementing/overriding a specific method that the util function could call, perhaps the `__str__` itself?

I admit not having given much thought about branched networks. I'd initially conceived of looping through the layers list to detect a merge layer and taking it from there. Alternately, perhaps the `__str__` could be built recursively through the layers, where the merge layer handles the specifics of the merge representation? But the latter would be moving away from implementing this as a utility function. 

I'm not sure how this would work for networks that branch out and don't merge later. I'm perhaps not sure how such networks are handled in Lasagne either. I suppose you'd have to use a function handle for each branch?

**Comment by ebenolson** [comment]
Have a look at the code of the [`get_all_layers`](https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/helper.py#L23) helper function for the current approach to branching layers. 

I was more wondering how you would lay out a complex network in text form. I did a bit of work on something similar a while back (https://github.com/Lasagne/Lasagne/issues/174), but I let a graph library do most of the job.

For simple linear networks, I don't really see the advantage over just printing a list. 
But I do think it would be nice to define `__str__` so instead of getting 

    <lasagne.layers.dnn.Conv2DDNNLayer at 0x7fa13df38e90>

we could have something like

    <Conv2DDNNLayer: filter_size=(3, 3), output_shape=(None, 11, 7, 7)>

**Comment by sidps** [comment]
I second that, it would indeed be excellent if we could have informative `__str__`s. Perhaps it would be good to describe a standard as well, if this is going to be the approach in future.

For textual representation of the network, I initially intended mimicking torch loosely, but that doesn't have to be (I'm not sure they're immediately comprehensible either). An alternative would be to implement something like `git log --graph`, but handling cycles may not be straightforward.

**Comment by sidps** [comment]
See [Recipes PR # 3](https://github.com/Lasagne/Recipes/pull/3).
